skbio/io/format/genbank.py

```diff
@@ -0,0 +1,468 @@
+|Yes   |Yes   |:mod:`skbio.sequence.Protein`                                  |
++------+------+---------------------------------------------------------------+
+
+Format Specification
+--------------------
+The following sections define the genbank formats.
+
+<http://www.ncbi.nlm.nih.gov/Sitemap/samplerecord.html>
+
+Feature Table Documentation:
+http://www.insdc.org/files/feature_table.html
+ftp://ftp.ncbi.nih.gov/genbank/docs/FTv10_3.html
+
+    The International Nucleotide Sequence Database Collaboration (INSDC)
+    between the DDBJ, EMBL, and GenBank.  These organisations all use the
+    same "Feature Table" layout in their plain text flat file formats.
+
+    However, the header and sequence sections of an EMBL file are very
+    different in layout to those produced by GenBank/DDBJ.
+
+Examples
+--------
+
+"""
+# ----------------------------------------------------------------------------
+# Copyright (c) 2013--, scikit-bio development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import range, zip
+from six.moves import zip_longest
+import re
+import textwrap
+
+import numpy as np
+import pandas as pd
+from datetime import datetime
+from functools import partial
+
+from skbio.io import create_format, GenbankFormatError
+from skbio.io.registry import FileSentinel
+from skbio.io.format._base import (_get_nth_sequence,
+                                   _parse_fasta_like_header,
+                                   _format_fasta_like_records, _line_generator,
+                                   _too_many_blanks)
+from skbio.util._misc import chunk_str
+from skbio.sequence import Sequence, DNA, RNA, Protein
+from pprint import pprint
+
+genbank = create_format('genbank')
+
+# date format in locus line of genbank record
+_TIME_FORMAT = '%d-%b-%Y'
+# This list is ordered
+# used to read and write genbank file.
+_HEADERS = ['LOCUS',
+            'DEFINITION',
+            'ACCESSION',
+            'VERSION',
+            'DBSOURCE',
+            'DBLINK',
+            'KEYWORDS',
+            'SOURCE',
+            'REFERENCE',
+            'COMMENT',
+            'FEATURES',
+            'ORIGIN']
+
+
+@genbank.sniffer()
+def _genbank_sniffer(fh):
+    # check the 1st line starts with 'LOCUS' and the last line ends with '//'
+    if _too_many_blanks(fh, 5):
+        return False, {}
+    try:
+        line = next(_line_generator(fh, skip_blanks=True))
+    except StopIteration:
+        return False, {}
+
+    if line.startswith(_HEADERS[0]):
+        for line in _line_generator(fh, skip_blanks=True):
+            if line == '//':
+                return True, {}
```

Comment:
It's still more work than needed. I think in the case of an incomplete record, the user would be fine with the reader raising an exception, having already identified it as `genbank` versus the registry saying it has no idea what the file is.